ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,249 @@ func (ps *prioritySampler) apply(spn *span) {
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Name != "":
+				rules = append(rules, NameServiceRule(v.Name, v.Service, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Name != "":
+				rules = append(rules, NameRule(v.Name, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+// sampleRate returns the sampling rate found in the DD_TRACE_SAMPLE_RATE environment variable. If it is invalid and
+// not within the 0-1 range, the default value of 0 is returned.
+func sampleRate() float64 {
+	defaultRate := math.NaN()
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return defaultRate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("ignoring DD_TRACE_SAMPLE_RATE: error: %v", err)
+		return defaultRate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("ignoring DD_TRACE_SAMPLE_RATE: out of range %f", r)
+	return defaultRate
+}
+
+// newRateLimiter returns a new rate limiter which allows r events per second.
+// The DD_TRACE_LIMIT environment variable may override r.
+func newRateLimiter(r float64) *rate.Limiter {
+	defaultLimiter := rate.NewLimiter(rate.Inf, 0)
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return defaultLimiter
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return defaultLimiter
+	}
+	if l < 0.0 {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is negative: %f", l)
+		return defaultLimiter
+	}
+	return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+}
+
+// apply uses the sampling rules to determine the sampling rate for the
+// provided span. If the rules don't match, and a default rate hasn't been
+// set using DD_TRACE_SAMPLE_RATE, then it returns false and the span is not
+// modified.
+func (rs *rulesSampler) apply(span *span) bool {
+	var matched bool
+	rate := rs.rate
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			matched = true
+			rate = rule.Rate
+			break
+		}
+	}
+	if !matched && math.IsNaN(rate) {
+		// no matching rule or global rate, so we want to fall back
+		// to priority sampling
+		return false
+	}
+
+	rs.applyRate(span, rate, time.Now())
+	return true
+}
+
+func (rs *rulesSampler) applyRate(span *span, rate float64, now time.Time) {
+	// rate sample
+	span.SetTag(keyRulesSamplerAppliedRate, rate)
+	if !sampledByRate(span.TraceID, rate) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return
+	}
+	// global rate limit and effective rate calculations
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	if d := now.Sub(rs.ts).Truncate(time.Second); d >= time.Second {
```

Comment:
This is way too complicated and "mathsy", with no explanations. It's probably just me, but I've been trying to understand this bit of code on every review I do and I can't. Some questions:

1. Can we get rid of the locks? It introduces a small chance for the rate being slightly off, but I think it should be negligible.
2. Can we test it better and document it so that a reader better understands what this does?

* `// update "previous rate" and reset` sounds more like code than an actual clarification of the logic and reasoning behind it.
* The same for `// global rate limit and effective rate calculations`.

Reply:
Maybe something like:
```go
// compute the effective rate of the rate limiter by averaging the rate for the previous second with the current
if d := now.Sub(rs.previousTime); d >= time.Second {
	// more than a second has passed
	if d-time.Second < time.Second && rs.seen > 0 {
                // no more than one second has passed, store the previous rate
		rs.previousRate = float64(rs.allowed) / float64(rs.seen)
	} else {
                // more than one second has passed, disregard older rate
		rs.previousRate = 0.0
	}
	rs.previousTime = now
	rs.allowed = 0
	rs.seen = 0
}
```
Note that I've removed `.Truncate(time.Second)` from the code.

Reply:
Good point. You can leave the lock, but please add the comments.